PageRank/markov_chain.py

```diff
@@ -848,7 +848,120 @@ def discuss_markov_prop(self, markov_chain_g):
+        )
+
+        markov_chain_g = MarkovChainGraph(markov_chain, enable_curved_double_arrows=True, layout="circular")
+        markov_chain_g.scale(1.1)
+        markov_chain_t_labels = markov_chain_g.get_transition_labels()
+
+        self.play(
+            FadeIn(markov_chain_g)   
+        )
+        self.wait()
+
+        base_ranking_values = [0.95, 0.75, 0.5, 0.25]
+        original_width = markov_chain_g.vertices[0].width
+        final_ranking = self.show_randomized_ranking(markov_chain_g, base_ranking_values)
+
+        
+
+        how_to_measure_importance = Text("How to Measure Relative Importance?", font="CMU Serif", weight=BOLD).scale(0.8)
+        how_to_measure_importance.move_to(title.get_center())
+        self.play(
+            *[
+            markov_chain_g.vertices[v].animate.scale_to_fit_width(original_width) for v in markov_chain.get_states()
+            ],
+            FadeOut(final_ranking),
+            ReplacementTransform(title, how_to_measure_importance),
+        )
+        self.wait()
+
+        markov_chain_sim = MarkovChainSimulator(
+            markov_chain, markov_chain_g, num_users=100
+        )
+        users = markov_chain_sim.get_users()
+
+        self.play(*[FadeIn(user) for user in users])
+        self.wait()
+
+        num_steps = 5
+        for _ in range(num_steps):
+            transition_map = markov_chain_sim.get_lagged_smooth_transition_animations()
+            self.play(
+                *[LaggedStart(*transition_map[i]) for i in markov_chain.get_states()]
+            )
+            self.wait()
+        self.wait()
+
+    def show_randomized_ranking(self, markov_chain_g, base_ranking_values):
+        original_markov_chain_nodes = [markov_chain_g.vertices[i].copy() for i in range(len(base_ranking_values))]
+        positions = [LEFT * 2.4, LEFT * 0.8, RIGHT * 0.8, RIGHT * 2.4]
+        gt_signs = [MathTex(">"), MathTex(">"), MathTex(">")]
+        for i, sign in enumerate(gt_signs):
+            gt_signs[i].move_to((positions[i] + positions[i + 1]) / 2)
+        num_iterations = 5
+        SHIFT_DOWN = DOWN * 3.2
+        for step in range(num_iterations):
+            print('Iteration', step)
+            current_ranking_values = self.generate_new_ranking(base_ranking_values)
+            current_ranking_map = self.get_ranking_map(current_ranking_values)
+            scaling_animations = []
+            for v, scaling in current_ranking_map.items():
+                scaling_animations.append(
+                    markov_chain_g.vertices[v].animate.scale_to_fit_width(scaling)
+                )
+            current_ranking = self.get_ranking(current_ranking_map)
+            ranking_animations = []
+            for i, v in enumerate(current_ranking):
+                if step != 0:
+                    ranking_animations.append(
+                        original_markov_chain_nodes[v].animate.move_to(positions[i] + SHIFT_DOWN)
```

Comment:
We can make this a circular movement along an arc so the movement is a bit clearer! I think transforms have an option to make it follow an arc path, but i don't quite remember. Also, I remember a Swap class, but just saw the source code and it is not implemented haha. But there are lots of options!